spectec/src/middlend/unthe.ml

```diff
@@ -0,0 +1,209 @@
+(*
+This transformation removes use of the ! operator from relations, by
+introducing fresh variables.
+
+An occurrence of !(e) will be replaced with a fresh variable x of the suitable
+type (and dimension), and a new condition e = ?(x) is added.
+
+This is an alternative to to how the Sideconditions pass handles the ! operator.
+If you need both, passes, run this one first.
+*)
+
+open Util
+open Source
+open Il.Ast
+
+(* Errors *)
+
+let error at msg = Source.error at "sideconditions" msg
+
+(* We pull out fresh variables and equating side conditions. *)
+
+type bind = (id * typ * iter list)
+type eqn = (bind * premise)
+type eqns = eqn list
+
+(* Fresh name generation *)
+
+let name i = "o" ^ string_of_int i (* no clash avoidance *)
+
+let fresh_id (n : int ref) : id =
+  let i = !n in
+  n := !n+1;
+  name i $ no_region
+
+(* If the expression (or premise) under iteration changes, we may be able to
+drop some variables from the iterexp *)
+
+let update_iterexp_vars (sets : Il.Free.sets) ((iter, vs) : iterexp) : iterexp =
+  (iter, List.filter (fun v -> Il.Free.Set.mem v.it sets.varid) vs)
+
+(* If a bind and premise is generated under an iteration, wrap them accordingly *)
+
+let under_iterexp (iter, vs) eqns : iterexp * eqns =
+   let new_vs = List.map (fun ((v, _, _), _) -> v) eqns in
+   let iterexp' = (iter, vs @ new_vs) in
+   let eqns' = List.map (fun ((v, t, is), pr) ->
+     let pr_iterexp = update_iterexp_vars (Il.Free.free_prem pr) (iter, vs @ new_vs) in
+     let pr' = IterPr (pr, pr_iterexp) $ no_region in
+     ((v, t, is@[iter]), pr')
+   ) eqns in
+   iterexp', eqns'
+
+
+(* Generic traversal helpers *)
+
+type 'a traversal = int ref -> 'a -> eqns * 'a
+type ('a, 'b) traversal_k = int ref -> 'a -> ('a -> 'b) -> eqns * 'b
+
+let phrase (t : 'a traversal) : ('a, 'b) note_phrase traversal
+  = fun n x -> let eqns, x' = t n x.it in eqns, x' $$ x.at % x.note
+
+let t_list (t : 'a traversal) : ('a list, 'b) traversal_k
+  = fun n xs k ->
+    let eqnss, xs' = List.split (List.map (t n) xs) in
+    List.concat eqnss, k xs'
+
+let unary (t : 'a traversal) : ('a, 'b) traversal_k =
+  fun n x k ->
+  let eqns, exp' = t n x in
+  eqns, k exp'
+
+let binary (t1 : 'a traversal) (t2 : 'b traversal) : ('a * 'b, 'c) traversal_k =
+  fun n (x1, x2) k ->
+  let eqns1, x1' = t1 n x1 in
+  let eqns2, x2' = t2 n x2 in
+  eqns1 @ eqns2, k (x1', x2')
+
+let ternary (t1 : 'a traversal) (t2 : 'b traversal) (t3 : 'c traversal) : ('a * 'b * 'c, 'd) traversal_k =
+  fun n (x1, x2, x3) k ->
+  let eqns1, x1' = t1 n x1 in
+  let eqns2, x2' = t2 n x2 in
+  let eqns3, x3' = t3 n x3 in
+  eqns1 @ eqns2 @ eqns3, k (x1', x2', x3')
+
+(* Expr traversal *)
+
+let rec t_exp n e : eqns * exp =
+  (* Descend first using t_exp2, and then see if we have to pull out the current expression *)
+  let eqns, e' = t_exp2 n e in
+  match e.it with
```

Comment:
@nomeata - Was this supposed to be `e'.it` instead of `e.it`?

Reply:
Oh, quite likely. We probably don’t have nested `the` so far. Thanks for spotting!